R/unit.R

```diff
@@ -145,258 +150,198 @@ calcStringMetric <- function(text) {
-        else
-            n <- length(data)
-        original <- data
-        length(data) <- n
-        n.o <- length(original)
-        if (n.o < n)
-            for (i in (n.o + 1L):n)
-                data[[i]] <- original[[(i - 1L) %% n.o + 1L]]
-    }
-    data
-}
-
-# Make sure that and "str*" and "grob*" units have data
-valid.data <- function(units, data) {
-    n <- length(units)
-    str.units <- stringUnit(units)
-    if (any(str.units))
-        for (i in (1L:n)[str.units])
-            if (!(length(data) >= i &&
-                  (is.character(data[[i]]) || is.language(data[[i]]))))
-                stop("no string supplied for 'strwidth/height' unit")
-    # Make sure that a grob has been specified
-    grob.units <- grobUnit(units)
-    if (any(grob.units))
-        for (i in (1L:n)[grob.units]) {
-            if (!(length(data) >= i &&
-                  (is.grob(data[[i]]) || inherits(data[[i]], "gPath") ||
-                   is.character(data[[i]]))))
-                stop("no 'grob' supplied for 'grobwidth/height' unit")
-            if (is.character(data[[i]]))
-                data[[i]] <- gPath(data[[i]])
-            if (inherits(data[[i]], "gPath"))
-                if (depth(data[[i]]) > 1)
-                    stop("'gPath' must have depth 1 in 'grobwidth/height' units")
-        }
-    # Make sure that where no data is required, the data is NULL
-    if (!all(sapply(data[!(str.units | grob.units)], is.null)))
-        stop("non-NULL value supplied for plain unit")
-    data
-}
+                     "mylines", "mychar", "mystrwidth", "mystrheight",
+                     "sum", "min", "max")
 
 valid.units <- function(units) {
   .Call(C_validUnits, units)
 }
 
+# Printing, formating, and coercion
+unitDesc <- function(x, format = FALSE, ...) {
+  amount <- if (format) format(x[[1]], ...) else x[[1]]
+  unit <- units[as.character(x[[3]])]
+  if (unit %in% c('sum', 'min', 'max')) {
+    paste0(if (amount == 1) '' else paste0(amount, '*'),
+           unit, '(', paste(lapply(x[[2]], unitDesc, format = format, ...), collapse = ', '), ')')
+  } else {
+    paste0(amount, unit)
+  }
+}
 as.character.unit <- function(x, ...) {
-  class(x) <- NULL
-  paste0(x, attr(x, "unit"))
+  vapply(as.unit(x), unitDesc, character(1))
 }
-
+as.double.unit <- function(x, ...) {
+  vapply(unclass(x), `[[`, numeric(1), 1L)
+}
+as.vector.unit <- as.double.unit
 format.unit <- function(x, ...) {
-    paste0(format(unclass(x), ...), attr(x, "unit"))
+  vapply(as.unit(x), unitDesc, character(1), format = TRUE, ...)
 }
+print.unit <- function(x, ...) {
+  print(as.character(x), quote = FALSE, ...)
+  invisible(x)
+}
+as.double.simpleUnit <- function(x, ...) as.double(unclass(x), ...)
+as.vector.simpleUnit <- function(x, ...) as.double(unclass(x), ...)
 
+is.unit <- function(x) inherits(x, 'unit')
+is.simpleUnit <- function(x) inherits(x, 'simpleUnit')
+identicalUnits <- function(x) .Call(C_conformingUnits, x)
+
+as.unit <- function(x) {
```

Comment:
Is this where "simpleUnit"s get converted to the standard "unit" ?  So all functions that work with units MUST call as.unit() ?

Reply:
I see some places in Ops.unit() checking explicitly for is.simpleUnit().  That looks like an easy opportunity to make a mistake.  Would it be much simpler to drop "simpleUnit"s altogether ?

Reply:
I see that "simpleUnits"s are entwined everywhere - presumably because the performance improvement is worth it (?).  Even so, the concern remains that we always have to check for "simpleUnit"s.  Is there a well-defined rule for when we need to check this, when we can assume that we are dealing with a standard "unit", etc ?

Reply:
See general comment. If the function can optimise on `simpleUnit` it should not convert, otherwise it should...

The performance is def worth it. On the C side you shouldn't worry about the type of unit as long as you use the getters. On the R side, it is only the low level stuff that should worry about what format the unit is in, but this was always the case as currently there are 3 very different unit types

Reply:
Yep.  Happy now thanks.